pandajedi/jedicore/JediDBProxy.py

```diff
@@ -13692,3 +13692,225 @@ def extendSandboxLifetime_JEDI(self, jedi_taskid, file_name):
+        methodName = self.getMethodName(comment)
+        methodName += " < sql={0} >".format(sql_query)
+        tmpLog = MsgWrapper(logger, methodName)
+        try:
+            self.conn.begin()
+            # sql to query
+            self.cur.execute(sql_query+comment, params_map)
+            taskIDs = self.cur.fetchall()
+            # sql to put the task in pending
+            sqlPDG = (  "UPDATE {0}.JEDI_Tasks "
+                        "SET lockedBy=NULL, lockedTime=NULL, "
+                            "status=:status, errorDialog=:err, "
+                            "modificationtime=CURRENT_DATE, oldStatus=status "
+                        "WHERE jediTaskID=:jediTaskID "
+                            "AND status IN ('ready','running','scouting') "
+                            "AND lockedBy IS NULL "
+                      ).format(jedi_config.db.schemaJEDI)
+            # loop over tasks
+            n_updated = 0
+            for jedi_taskid in taskIDs:
+                varMap = {}
+                varMap[':jediTaskID'] = jedi_taskid
+                varMap[':err'] = reason
+                varMap[':status'] = 'pending'
+                self.cur.execute(sqlPDG+comment, varMap)
+                nRow = self.cur.rowcount
+                if nRow == 1:
+                    self.record_task_status_change(jedi_taskid)
+                    n_updated += 1
+                elif nRow > 1:
+                    tmpLog.error('updated {0} rows with same jediTaskID'.format(nRow))
+            if not self._commit():
+                raise RuntimeError('Commit error')
+            tmpLog.debug('done with {0} rows'.format(n_updated))
+            # return
+            return n_updated
+        except Exception:
+            # roll back
+            self._rollback()
+            # error
+            self.dumpErrorMessage(tmpLog)
+            return None
+
+
+    # get  datasets of input and lib, to update data locality records
+    def get_tasks_inputdatasets_JEDI(self, vo):
+        comment = ' /* JediDBProxy.get_tasks_inputdatasets_JEDI */'
+        methodName = self.getMethodName(comment)
+        # last update time
+        methodName += " <vo={0}>".format(vo)
+        tmpLog = MsgWrapper(logger, methodName)
+        tmpLog.debug('start')
+        now_ts = datetime.datetime.utcnow()
+        try:
+            retVal = None
+            # sql to get all jediTaskID and datasetID of input
+            sql = ( "SELECT tabT.jediTaskID,datasetID, tabD.datasetName "
+                    "FROM {0}.JEDI_Tasks tabT,{0}.JEDI_Datasets tabD,{0}.JEDI_AUX_Status_MinTaskID tabA "
+                    "WHERE tabT.status=tabA.status AND tabT.jediTaskID>=tabA.min_jediTaskID AND tabT.jediTaskID=tabD.jediTaskID "
+                        "AND tabT.vo=:vo AND tabT.modificationTime>=:modificationTime "
+                        "AND tabD.type IN ('input') AND tabD.masterID IS NULL "
+                    ).format(jedi_config.db.schemaJEDI)
```

Comment:
Thanks. I modified the query as your suggestion, and I also add another condition that modificationtime within last 60 days, not sure if it's reasonable. Now it runs better.

As to adding new column to task table, it sounds a bigger change and I will leave it as last resort.